keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -1249,6 +1256,181 @@ of `processing latency` used for determining the weight of the queue may
+This means that some costs may be overestimated, but if we resaonably hash
+requests originating by system components, that seems acceptable.
+The above can be achieved by hashing each resource type to a separate set of
+buckets, and within a resource type hashing (namespace, name) as simple as:
+```
+  hash(ns, name) = 0                              if namespace == "" & name == ""
+  hash(ns, name) = 1 + hash(namespace)%A          if name == ""
+  hash(ns, name) = 1 + A + hash(namespace/name)%B otherwise
+```
+For small enough A and B (e.g. A=3, B=6), the representation should have less
+than 1000 entries, so it would be small enough to make periodic updates in etcd
+reasonable.
+
+We can optimize amount of data written to etcd by frequently (say once per
+second) checking what changed, but writing rarely (say once per minute) or
+if values in some buckets significantly increased.
+The above algorithm would allow us to avoid some more complicated time-smearing,
+as whenever something quickly grows we report it, but we don't immediately
+downscale which is a way to somehow incorporate a history.
+
+However, we will treat the above as a feasibility proof.  We will just start
+with the simplest apprach of treating each kube-apiserver independently.
+We will implement the above (i.e. knowledge sharing between kube-apiserver),
+if the independence assumption will not work good enough.
+The above description shows that it won't result in almost any wasted work
+if the code will be well structured.
+
+##### Cost of the watch event
+
+We assumed above that the cost of processing every watch event is equal.
+However, in practice the cost associated with sending an event consists
+of two main parts:
+- the cost of going through event change logic
+- the cost of processing the even object (e.g. deserialization or sending data
+  over network)
+The first one is close to equal independently of the event, the second one
+is more proportional to the size of the object.
+However, estimating size of the is hard to predict for PATCH or DELETE requests.
+Additionally, even for POST or PUT requests where we could potentially
+estimate it based on size of the body of the requests, we may not yet have
+access to it when we need to make a decision.
+
+One way to estimate it would be to keep a running average of watch event size
+per bucket.  While it won't give us an accurate estimate, it should amortize
+well over time.
+
+Obviously some normalization will be needed here, but it's impossible to assess
+it on paper, and we are leaving this for the implementation and tuning phase
+to figure out the details.
+
+##### Dispatching the request
+
+We described how we can estimate the cost of the request associated with
+watch events triggerred by it.  But we didn't yet said how this translates
+to dispatching the request.
+
+First of all, we need to decide how to translate the estimated cost to the
+`width` of the request and its latency.  Second, we need to introduce the
```

Comment:
I'd vote for adding latency to the end of the request rather than adding width, I think that tracks reality a little better?

Reply:
To be clear - what I recommend at the end is come combination. In majority of cases it would just be translating to latency. But with large number of watchers, it will be translating to both width and latency.

Reply:
OK I'll comment there if I have something else to say about it.